add blog post for constructors in golang (#213)
deliveroo/deliveroo.engineering: _posts/2019-12-14-increase-the-reliability-of-a-golang-codebase-with-object-constructors.md

---
layout: post
title:  "Increase the Reliability of a Golang Codebase with Object Constructors"
authors:
  - "Tugberk Ugurlu"
excerpt: >
  One of the limitation of Go programming language is the lack of built-in object constructor 

>   support. In this post, we will see how this have a negative impact on the code we write and 

Reviewer: `have` should be `can have` or `has`.

One of these limitations I have come across is not [having any built-in constructor support](https://twitter.com/tourismgeek/status/1074325233220374528).

I came across this limitation before using Golang in an anger and I was anticipating this to be 
a bit of a limitation. At the same time, I was also being open-minded. After seeing a few of the 
problems that lack of constructors caused when I started getting some experience under my belt 
with Go, I am keen to document what I think is the solution here is. 

> I must give credit to [John Arundel](https://twitter.com/bitfield). Thanks to [the discussion we have had on Twitter](https://twitter.com/bitfield/status/1074637347193581568), I am able 
to express a solution to this problem here which is based on [what John made me aware of first](https://twitter.com/bitfield/status/1074682570389028866).

Now, when I say constructors in the title of the post here, I must confess that it’s a bit of a 
lie as I don’t see a way of having pure object constructors like we have with C# or Java in Go 

> without changes in the language itself. However, we can workaround the lack of constructors in 

Reviewer 2: "we can work around" or "we can have a workaround", I think

> without changes in the language itself. However, we can workaround the lack of constructors in 

Author: ```suggestion
without changes in the language itself. However, we can work around the lack of constructors in 
```

Go by leveraging some other aspects of the language such as package scoping and interfaces and 
essentially [adopt the factory method pattern](https://en.wikipedia.org/wiki/Factory_method_pattern).

Let’s first touch on these two aspects of Go and we will then see how we can take advantage of 
these aspects to make our code more robust and protect against unexpected consumptions in the 
feature.

## Package Scoping
Go doesn’t have access modifiers such as private, internal or public per se. However, you can 
influence whether a type should be internal to a package or should be exposed through naming in 
Go. When your type is named by starting with a lowercase letter, it will only be available 
within the package itself. This rule also applies to the functions, and members of the types 
such as fields and methods.

For example, the below code sample does not compile successfully in Go:

`singers/jazzsinger.go` file:

```golang
package singers

type jazzSinger struct {
}

func (jazzSinger) Sing() string {
	return "Des yeux qui font baisser les miens"
}
```

`main.go` file:

```golang
package main

import (
	"fmt"

>             "github.com/tugberkugurlu/go-package-scope/singers"

Reviewer 3: mixed indentation?

>             "github.com/tugberkugurlu/go-package-scope/singers"

Author: ```suggestion
        "github.com/tugberkugurlu/go-package-scope/singers"
```

)

func main() {
	s := singers.jazzSinger{}
	fmt.Println(s.Sing())
}
```

If we were to run this code, we would get the below error:

```
➜  go-package-scope go run main.go
# command-line-arguments
./main.go:9:7: cannot refer to unexported name singers.jazzSinger
./main.go:9:7: undefined: singers.jazzSinger
```

This sort of demonstrates how package scoping works in Go. You can learn more about packages in 
Go from [Uday's great article on this topic](https://medium.com/rungo/everything-you-need-to-know-about-packages-in-go-b8bac62b74cc) 
but this should be enough for us to get going for our example.

## Interfaces

> Let's now look at interfaces in Go, which act very similar to what you would expect them to be. However, the way you implement interfaces is very different to how you would do in C#, Java or TypeScript. The main difference is that you don’t explicitly declare that a struct implements an interface in Go. A struct is considered as an implementation of an interface by the compiler as long as it implements all the methods within it. Let’s look at the following example:

Reviewer 3: `as long as it implements all the methods within it`

Should we mention that all of the method signatures have to match too?

Author: good suggestion, I will add it 👍

> Let's now look at interfaces in Go, which act very similar to what you would expect them to be. However, the way you implement interfaces is very different to how you would do in C#, Java or TypeScript. The main difference is that you don’t explicitly declare that a struct implements an interface in Go. A struct is considered as an implementation of an interface by the compiler as long as it implements all the methods within it. Let’s look at the following example:

Author: ```suggestion
Let's now look at interfaces in Go, which act very similar to what you would expect them to be. However, the way you implement interfaces is very different to how you would do in C#, Java or TypeScript. The main difference is that you don’t explicitly declare that a struct implements an interface in Go. A struct is considered as an implementation of an interface by the compiler as long as it implements all the methods within it with matching signatures. Let’s look at the following example:
```

goroutine 1 [running]:
github.com/tugberkugurlu/go-package-scope/singers.jazzSinger.Sing(0x0, 0x0, 0x1010095, 0xc00008e030)
	/Users/tugberkugurlu/go/src/github.com/tugberkugurlu/go-package-scope/singers/jazzsinger.go:31 +0x22
main.singToConsole(0x10d75a0, 0xc00008e030)
	/Users/tugberkugurlu/go/src/github.com/tugberkugurlu/go-package-scope/main.go:14 +0x35
main.main()
	/Users/tugberkugurlu/go/src/github.com/tugberkugurlu/go-package-scope/main.go:10 +0x5c
```

There isn't a solution available in Go as far as I am aware which would allow us to fail for these cases during compilation. However, thanks to the dedicated constructor for this object, we can explicitly signal potential construction errors by returning multiple values from the function call:

```golang
func NewJazzSinger(lyrics LyricsProvider) (Singer, error) {
	if lyrics == nil {
		return nil, errors.New("lyrics cannot be nil")
	}
	return jazzSinger{
		lyrics: lyrics,
	}, nil
}
```

At the time of consumption, it becomes very explicit to deal with returned result:

```golang
s, err := singers.NewJazzSinger(nil)
if err != nil {
	log.Fatal(err)
}
// ...
```

> ## Do I Really Need Interfaces For This?

Reviewer 2: I'm missing some `# Conclusions` section here :)

Author: yep, will add one now :)

Author: added here: https://github.com/deliveroo/deliveroo.engineering/pull/213/commits/2ece97e246f179ca391ea114e03bac52e1ea5e0b pls lmk what you think?

type by forcing them to choose between a `class` and `struct`.

In Go, you can make this safer through the use of the constructor pattern as well, by ensuring that your struct is not allowed to be constructed directly and you controlling how the initialized value should be flowed.

```golang
package singers

type Singer interface {
	Sing()  string
	Count() int
}

type jazzSinger struct {
	count int
}

func (j *jazzSinger) Sing() string {
	j.count++
	return "Des yeux qui font baisser les miens"
}

func (j *jazzSinger) Count() string {
	return j.count
}

func NewJazzSinger() Singer {
	return &jazzSinger{}
}
```

The consumer of this type needs to construct it through `NewJazzSinger` function here, which is making the decision to flow the type as a pointer because it needs to be able to mutate its own state as it's being used.

```golang
package main

import (
	"fmt"
	"github.com/tugberkugurlu/go-package-scope/singers"
)

func main() {
	s := singers.NewJazzSinger()
	singToConsole(s)
	fmt.Println(s.Count())
	singToConsole(s)
	fmt.Println(s.Count())
}

func singToConsole(singer singers.Singer) {
	fmt.Println(singer.Sing())
}
```

## Conclusion

> Modeling your domain is hard and it's also harder if you have rich models, which hold a state. Go programming language may not give you all the tools directly to model your domain in a rich way as some other programing languages provides, it's still possible to 

Author: ```suggestion
Modeling your domain is hard and it's even harder if you have rich models which hold a mutable state along with explicit behaviours. Go programming language may not give you all the tools directly to model your domain in a rich way as some other programing languages provides. However, it's still possible to 
```

make it work for some cases by adopting some usage principles. Constructor pattern is one of them, and it has been one of the most 

> useful ones for me since I can confidently encapsulate the logic for my model within a package as well as its construction.

Author: ```suggestion
useful ones for me since I can confidently encapsulate the initialisation logic of my model by enforcing state validity within a package as well as its construction.
```

> useful ones for me since I can confidently encapsulate the initialisation logic of my model by enforcing state validity within a package as well as its construction.

Author: ```suggestion
useful ones for me since I can confidently encapsulate the initialisation logic of my model by enforcing state validity within a package scope.
```
